src/program.cpp

```diff
@@ -758,176 +443,68 @@ void program::debug_print(instruction_ref ins) const
-{
-    os << "migraphx::shape{migraphx::shape::" << s.type_string();
-    os << ", {" << to_string_range(s.lens()) << "}";
-    if(not s.standard())
-        os << ", {" << to_string_range(s.strides()) << "}";
-    os << "}";
-}
-
-void program::print_cpp(std::ostream& os) const
-{
-    os << "migraphx::program p;" << std::endl;
-    // cppcheck-suppress variableScope
-    unsigned long seed = 0;
-    print_program(*this, [&](auto ins, const auto& names) {
-        auto op = cpp_op_var(names.at(ins), ins);
-        if(ins->name().front() != '@')
-        {
-            os << "migraphx::op::" << ins->name() << " " << op << ";" << std::endl;
-            print_op_attributes(os, op, ins->get_operator());
-        }
-        os << "auto " << cpp_var_name(names.at(ins)) << " = ";
-        if(ins->name() == "@literal")
-        {
-            os << "p.add_literal(";
-            bool use_abs = false;
-            ins->get_literal().visit([&](auto v) {
-                use_abs = std::none_of(v.begin(), v.end(), [](auto x) { return x < 0; });
-            });
-            if(use_abs)
-                os << "migraphx::abs(";
-            os << "migraphx::generate_literal(";
-            print_cpp_shape(os, ins->get_shape());
-            os << ", " << seed << ")";
-            if(use_abs)
-                os << ")";
-            os << ");" << std::endl;
-            seed++;
-        }
-        else if(ins->name() == "@param")
-        {
-            std::string name = any_cast<builtin::param>(ins->get_operator()).parameter;
-            os << "p.add_parameter(" << enclose_name(name) << ",";
-            print_cpp_shape(os, ins->get_shape());
-            os << ");" << std::endl;
-        }
-        else
-        {
-            os << "p.add_instruction(" << op;
-            for(auto input : ins->inputs())
-            {
-                os << ", " << cpp_var_name(names.at(input));
-            }
-            os << ");" << std::endl;
-        }
-    });
-}
 
-void program::dry_run(std::unordered_map<std::string, argument> params) const
-{
-    auto& ctx = this->impl->ctx;
-    generic_eval(*this, ctx, std::move(params), [](auto&&...) { return argument{}; });
+    return &impl->modules["main"];
 }
 
-void program::annotate(std::ostream& os, std::function<void(instruction_ref)> a) const
+const module* program::get_main_module() const
 {
-    print_program(*this, [&](auto ins, const auto& names) {
-        print_instruction(os, ins, names);
-        a(ins);
-        os << std::endl;
-    });
+    assert(contains(impl->modules, "main"));
+    return &impl->modules["main"];
 }
 
 program& program::sort()
 {
-    fix([&](auto self, auto ins) {
-        this->move_instruction(ins, this->begin());
-        for(auto child : ins->inputs())
-            self(child);
-    })(std::prev(this->end()));
-    assert(this->validate() == this->end());
+    assert(contains(impl->modules, "main"));
+    impl->modules["main"].sort();
```

Comment:
I think `program::sort()` can be removed.

Reply:
There are several locations that the program.sort() is called. Change all of them to prog.get_main_module()->sort()?